File: python/fluorophores/Fluorophore.py
from collections import namedtuple
import numpy as np

Properties = namedtuple("Properties", ["brightness", "pos"])

class Fluorophore:
    def __init__(self, pos=[0,0,0], brightness=1000):
        self.pos = pos  # nm
        self.brightness = brightness  # kHz
        self.remainingphotons = np.inf
        # Dummy properties for compatibility with collection
        self.numberOfFluorophores = 1
        self.allbleached = False

    @property
    def pos(self):
        return self._pos
    
    @pos.setter
    def pos(self, val):
        if not isinstance(val, np.ndarray):
            self._pos = np.array(val, dtype=float)
        else:
            self._pos = val
    
    def intensity(self, I0, dwelltime, time, phfac):
        # if props is not None:  # For performance: avoid property access
        #     brightness = props.brightness
        # else:
        brightness = self.brightness
        
        return brightness * I0 * dwelltime #* phfac
    
    def photons(self, I0, *args):
        return np.random.poisson(self.intensity(I0, *args))
    
    def reset(self, time=None):
        pass
    
    def position(self, time):
        pos = self.pos
        
        return pos, True
    
    # Dummy functions for compatibility with collection
    def updateonoff(self, time):
        pass
    
    def getproperties(self):
        return Properties(brightness=self.brightness, pos=self.pos)

File: MATLAB/fluorophores/Fluorophore.m
classdef Fluorophore<handle
    properties
        pos=[0, 0, 0]; %nm
        brightness=1000; %kHz;
        remainingphotons=inf;
        %dummy properties for compatibility with collection
        numberOfFluorophores=1;
        allbleached=false;
    end
    methods
        function obj=Fluorophore(args)
            arguments
                args.pos=[0,0,0];
                args.brightness=1000;
            end
            obj.pos=args.pos;
            obj.brightness=args.brightness;
        end
        function Io=intensity(obj,I0,dwelltime, time,phfac)
            % if nargin>=6 && ~isempty(props) %for performance: avoid property access
                % brightness=props.brightness;
            % else
                brightness=obj.brightness;
            % end
            
            Io=(brightness)*I0*dwelltime;%.*phfac;
        end
        function ph=photons(obj,I0,varargin)
            ph=poissrnd(obj.intensity(I0,varargin{:}));
        end
        function reset(obj,time)
        end
        function [posout,isactive]=position(obj,time)
            pos=obj.pos;
            posout=pos;
            isactive=true;
        end

        %dummy functions for compatibility with collection
        function updateonoff(obj,time)
        end
        function props=getproperties(obj)
            props.brightness=obj.brightness;
            props.pos=obj.pos;
        end
    end
end
